azure/services/privateendpoints/spec.go

```diff
@@ -82,55 +84,182 @@ func (s *PrivateEndpointSpec) Parameters(ctx context.Context, existing interface
+		}
 
-	if s.RequestMessage != "" {
-		privateLinkServiceConnection[0].PrivateLinkServiceConnectionProperties.RequestMessage = pointer.String(s.RequestMessage)
+		if privateLinkServiceConnection.RequestMessage != "" {
+			linkServiceConnection.PrivateLinkServiceConnectionProperties.RequestMessage = pointer.String(privateLinkServiceConnection.RequestMessage)
+		}
+		privateLinkServiceConnections = append(privateLinkServiceConnections, linkServiceConnection)
 	}
 
 	if s.ManualApproval {
-		privateEndpointProperties.ManualPrivateLinkServiceConnections = &privateLinkServiceConnection
+		privateEndpointProperties.ManualPrivateLinkServiceConnections = &privateLinkServiceConnections
+		privateEndpointProperties.PrivateLinkServiceConnections = &[]network.PrivateLinkServiceConnection{}
 	} else {
-		privateEndpointProperties.PrivateLinkServiceConnections = &privateLinkServiceConnection
+		privateEndpointProperties.PrivateLinkServiceConnections = &privateLinkServiceConnections
+		privateEndpointProperties.ManualPrivateLinkServiceConnections = &[]network.PrivateLinkServiceConnection{}
 	}
 
-	if s.ApplicationSecurityGroup != "" {
-		privateEndpointProperties.ApplicationSecurityGroups = &[]network.ApplicationSecurityGroup{
-			{
-				ID: pointer.String(s.ApplicationSecurityGroup),
-			},
-		}
+	applicationSecurityGroups := make([]network.ApplicationSecurityGroup, 0, len(s.ApplicationSecurityGroups))
+
+	for _, applicationSecurityGroup := range s.ApplicationSecurityGroups {
+		applicationSecurityGroups = append(applicationSecurityGroups, network.ApplicationSecurityGroup{
+			ID: pointer.String(applicationSecurityGroup),
+		})
 	}
 
-	return network.PrivateEndpoint{
-		Location:                  pointer.String(s.Location),
+	privateEndpointProperties.ApplicationSecurityGroups = &applicationSecurityGroups
+
+	newPrivateEndpoint := network.PrivateEndpoint{
+		Name:                      pointer.String(s.Name),
 		PrivateEndpointProperties: &privateEndpointProperties,
 		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
 			ClusterName: s.ClusterName,
 			Lifecycle:   infrav1.ResourceLifecycleOwned,
 			Name:        pointer.String(s.Name),
 			Additional:  s.AdditionalTags,
 		})),
-	}, nil
+	}
+
+	if s.Location != "" {
+		newPrivateEndpoint.Location = pointer.String(s.Location)
+	}
+
+	if existing != nil {
+		existingPE, ok := existing.(network.PrivateEndpoint)
+		if !ok {
+			return nil, errors.Errorf("%T is not a network.PrivateEndpoint", existing)
+		}
+
+		ps := existingPE.ProvisioningState
+		if string(ps) != string(infrav1.Canceled) && string(ps) != string(infrav1.Failed) && string(ps) != string(infrav1.Succeeded) {
+			return nil, azure.WithTransientError(errors.Errorf("Unable to update existing private endpoint in non-terminal state. Service Endpoint must be in one of the following provisioning states: Canceled, Failed, or Succeeded. Actual state: %s", ps), 20*time.Second)
+		}
+
+		normalizedExistingPE := network.PrivateEndpoint{
```

Comment:
Let's add an e2e test that changes these values if that wouldn't be too gnarly to set up.

Reply:
That might be a bit tricky as we need a real service endpoint available, otherwise the private endpoint creation will fail with something like:
```
privateendpoints failed to create or update. err: failed to create resource aneci-test-capz/test (service: privateendpoints): network.PrivateEndpointsClient#CreateOrUpdate: Failure sending request: StatusCode=404 -- Original Error: Code="ResourceNotFound" Message="Call to Microsoft.Storage/storageAccounts failed. Error message: The Resource 'Microsoft.Storage/storageAccounts/testaneci1' under resource group 'k8saneci-test-private-endpoint' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix
```